packages/foam-core/src/markdown-provider.ts

```diff
@@ -6,88 +6,106 @@ import { parse as parseYAML } from 'yaml';
 import { ID } from './types';
+import { ParserPlugin } from './plugins';
 
-let processor: unified.Processor | null = null;
-
-function parse(markdown: string): Node {
-  processor =
-    processor ||
-    unified()
-      .use(markdownParse, { gfm: true })
-      .use(frontmatterPlugin, ['yaml'])
-      .use(wikiLinkPlugin);
-  return processor.parse(markdown);
-}
-
-export function createNoteFromMarkdown(
-  uri: string,
-  markdown: string,
-  eol: string
-): Note {
-  const tree = parse(markdown);
-  let title: string | null = null;
-
-  visit(tree, node => {
-    if (node.type === 'heading' && node.depth === 1) {
-      title = ((node as Parent)!.children[0].value as string) || title;
-    }
-    return title === null ? CONTINUE : EXIT;
-  });
-
-  const links: NoteLink[] = [];
-  const linkDefinitions: NoteLinkDefinition[] = [];
-  let frontmatter: any = {};
-  let start: Point = { line: 1, column: 1, offset: 0 }; // start position of the note
-  visit(tree, node => {
+const yamlPlugin: ParserPlugin = {
+  visit: (node: Node, note: Note): void => {
     if (node.type === 'yaml') {
-      frontmatter = parseYAML(node.value as string) ?? {}; // parseYAML returns null if the frontmatter is empty
+      note.properties = parseYAML(node.value as string) ?? {};
+      // Give precendence to the title from the frontmatter if it exists
+      note.title = note.properties.title ?? note.title;
       // Update the start position of the note by exluding the metadata
-      start = {
+      note.source.contentStart = {
         line: node.position!.end.line! + 1,
         column: 1,
         offset: node.position!.end.offset! + 1,
       };
     }
+  },
+};
+
+const titlePlugin: ParserPlugin = {
+  visit: (node: Node, note: Note): void => {
+    if (note.title == null && node.type === 'heading' && node.depth === 1) {
+      note.title =
+        ((node as Parent)!.children[0].value as string) || note.title;
+    }
+  },
+};
 
+const wikilinkPlugin: ParserPlugin = {
+  visit: (node: Node, note: Note): void => {
     if (node.type === 'wikiLink') {
-      links.push({
+      note.links.push({
         type: 'wikilink',
         slug: node.value as string,
         position: node.position!,
       });
     }
+  },
+};
 
+const definitionsPlugin: ParserPlugin = {
+  visit: (node: Node, note: Note): void => {
     if (node.type === 'definition') {
-      linkDefinitions.push({
+      note.definitions.push({
         label: node.label as string,
         url: node.url as string,
         title: node.title as string,
         position: node.position,
       });
     }
-  });
-
-  // Give precendence to the title from the frontmatter if it exists
-  title = frontmatter.title ?? title;
+  },
+};
+
+export function createMarkdownParser(extraPlugins: ParserPlugin[]): NoteParser {
+  const parser = unified()
+    .use(markdownParse, { gfm: true })
+    .use(frontmatterPlugin, ['yaml'])
+    .use(wikiLinkPlugin);
+  // plugin.onDidInitializeParser(parser)
```

Comment:
nit: Remove?

Reply:
yeah, probably. The reason why it's there is because before implementing everything I wanted to highlight the possible extension points. So I am actually leaning more towards implementing that rather than removing it.
But I agree that this half baked code doesn't help (also, I realize I haven't documented this enough, will do in a bit)
Thoughts?